mypy/build.py

```diff
@@ -1341,7 +1340,20 @@ def get_cache_names(id: str, path: str, options: Options) -> tuple[str, str, str
         data_suffix = ".data.ff"
     else:
         data_suffix = ".data.json"
-    return (prefix + ".meta.json", prefix + data_suffix, deps_json)
+    return prefix + ".meta.json", prefix + data_suffix, deps_json
+
+
+def options_snapshot(id: str, manager: BuildManager) -> dict[str, object]:
+    """Make compact snapshot of options for a module.
+
+    Separately store only the options we may compare individually, and take a hash
+    of everything else. If --debug-cache is specified, fall back to full snapshot.
+    """
+    snapshot = manager.options.clone_for_module(id).select_options_affecting_cache()
```

Comment:
Random thing I noticed (no need to fix here, but could be useful to fix eventually if I'm right): `clone_for_module` does a bunch of caching to avoid repeated computation, but `select_options_affecting_cache` recomputes everything on each call. I think a similar caching strategy would work for it, which might improve options processing performance. If you agree with this, I can create an issue about this. The caching should probably also cover `hash_digest(json_dumps(...)``.